internal/controller/runner_controller.go

```diff
@@ -0,0 +1,354 @@
+			Namespace: namespace,
+		},
+		Spec: garmoperatorv1alpha1.RunnerSpec{},
+	}
+	err := r.Create(ctx, runnerObj)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.ensureFinalizer(ctx, runnerObj); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if _, err := r.updateRunnerStatus(ctx, runnerObj, garmRunner); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *RunnerReconciler) reconcileDelete(ctx context.Context, runnerClient garmClient.InstanceClient, garmRunner *params.Instance) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	if garmRunner != nil {
+		log.Info("Deleting Runner in Garm", "Runner Name", garmRunner.Name)
+		err := runnerClient.DeleteInstance(instances.NewDeleteInstanceParams().WithInstanceName(garmRunner.Name))
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *RunnerReconciler) getGarmRunnerInstance(client garmClient.InstanceClient, name string) (*params.Instance, error) {
+	// Problem: req.Name is not always name of garm instance like road-runner-k8s-EyCKa8uPE1to, but name of runner CR from cache which is lowercase road-runner-k8s-eycka8upe1to => GetInstanceByName can match runner in garm with name of CR
+	garmRunner, err := client.GetInstanceByName(instances.NewGetInstanceParams().WithInstanceName(name))
+	if err != nil && garmClient.IsNotFoundError(err) {
+		return nil, nil
+	} else if err != nil {
+		return nil, err
+	}
+	return &garmRunner.Payload, err
+}
+
+func (r *RunnerReconciler) ensureFinalizer(ctx context.Context, runner *garmoperatorv1alpha1.Runner) error {
+	if !controllerutil.ContainsFinalizer(runner, key.RunnerFinalizerName) {
+		controllerutil.AddFinalizer(runner, key.RunnerFinalizerName)
+		return r.Update(ctx, runner)
+	}
+	return nil
+}
+
+func (r *RunnerReconciler) updateRunnerStatus(ctx context.Context, runner *garmoperatorv1alpha1.Runner, garmRunner *params.Instance) (ctrl.Result, error) {
+	if garmRunner == nil {
+		return ctrl.Result{}, nil
+	}
+
+	log := log.FromContext(ctx)
+	log.Info("Update runner status...")
+
+	poolName := garmRunner.PoolID
+	pools := &garmoperatorv1alpha1.PoolList{}
+	err := r.List(ctx, pools)
+	if err == nil {
+		pools.FilterByFields(garmoperatorv1alpha1.MatchesID(garmRunner.PoolID))
+
+		if len(pools.Items) > 0 {
+			poolName = pools.Items[0].Name
+		}
+	}
+
+	runner.Status.ID = garmRunner.ID
+	runner.Status.ProviderID = garmRunner.ProviderID
+	runner.Status.AgentID = garmRunner.AgentID
+	runner.Status.Name = garmRunner.Name
+	runner.Status.OSType = garmRunner.OSType
+	runner.Status.OSName = garmRunner.OSName
+	runner.Status.OSVersion = garmRunner.OSVersion
+	runner.Status.OSArch = garmRunner.OSArch
+	runner.Status.Addresses = garmRunner.Addresses
+	runner.Status.Status = garmRunner.Status
+	runner.Status.InstanceStatus = garmRunner.RunnerStatus
+	runner.Status.PoolID = poolName
+	runner.Status.ProviderFault = garmRunner.ProviderFault
```

Comment:
can we convert the byte-array to a string (to get more readable error messages):

this is what i currently get

```yaml
apiVersion: garm-operator.mercedes-benz.com/v1alpha1
kind: Runner
metadata:
  creationTimestamp: "2023-12-08T12:58:05Z"
  finalizers:
  - garm-operator.mercedes-benz.com/runner
  generation: 1
  name: garm-xn2dldt5ae7r
  namespace: garm-operator-system
  resourceVersion: "35165"
  uid: a2b34960-7fd9-425a-9eb9-064ffa7252fc
spec: {}
status:
  agentId: 0
  githubRunnerGroup: ""
  id: 01cc1bd6-e0e8-45f2-ab16-82a8d82aebde
  instanceStatus: pending
  name: garm-xN2dldT5aE7R
  osArch: amd64
  osType: linux
  poolId: k8s-pool
  providerFault: Y3JlYXRpbmcgaW5zdGFuY2U6IHByb3ZpZGVyIGJpbmFyeSAvb3B0L2dhcm0vYmluL2dhcm0tcHJvdmlkZXItazhzIHJldHVybmVkIGVycm9yOiBwcm92aWRlciBiaW5hcnkgZmFpbGVkIHdpdGggc3Rkb3V0OiA7IHN0ZGVycjogMjAyMy8xMi8wOCAxMjo1ODo1OCBmYWlsZWQgdG8gcnVuIGNvbW1hbmQ6IGZhaWxlZCB0byBjcmVhdGUgaW5zdGFuY2UgaW4gcHJvdmlkZXI6IGVuc3VyaW5nIHJ1bm5lciBuYW1lc3BhY2UgcnVubmVyIGZhaWxlZDogbmFtZXNwYWNlcyAicnVubmVyIiBpcyBmb3JiaWRkZW46IFVzZXIgInN5c3RlbTpzZXJ2aWNlYWNjb3VudDpnYXJtLXNlcnZlcjpnYXJtLXNlcnZlciIgY2Fubm90IGdldCByZXNvdXJjZSAibmFtZXNwYWNlcyIgaW4gQVBJIGdyb3VwICIiIGluIHRoZSBuYW1lc3BhY2UgInJ1bm5lciIKOiBleGl0IHN0YXR1cyAx
  status: error
```

which means:

```
echo "Y3JlYXRpbmcgaW5zdGFuY2U6IHByb3ZpZGVyIGJpbmFyeSAvb3B0L2dhcm0vYmluL2dhcm0tcHJvdmlkZXItazhzIHJldHVybmVkIGVycm9yOiBwcm92aWRlciBiaW5hcnkgZmFpbGVkIHdpdGggc3Rkb3V0OiA7IHN0ZGVycjogMjAyMy8xMi8wOCAxMjo1ODozOCBmYWlsZWQgdG8gcnVuIGNvbW1hbmQ6IGZhaWxlZCB0byBjcmVhdGUgaW5zdGFuY2UgaW4gcHJvdmlkZXI6IGVuc3VyaW5nIHJ1bm5lciBuYW1lc3BhY2UgcnVubmVyIGZhaWxlZDogbmFtZXNwYWNlcyAicnVubmVyIiBpcyBmb3JiaWRkZW46IFVzZXIgInN5c3RlbhcnkgZmFpbGVkIHdpdGggc3Rkb3V0OiTpzZXJ2aWNlYWNjb3VudDpnYXJtLXNlcnZlcjpnYXJtLXNlcnZlciIgY2Fubm90IGdldCByZXNvdXJjZSAibmFtZXNwYWNlcyIgaW4gQVBJIGdyb3VwICIiIGluIHRoZSBuYW1lHJ1bm5lciBuYW1lc3BhY2UgcnVubmVyc3BhY2UgInJ1bm5lciIKOiBleGl0IHN0YXR1cyAx" | base64 -d                                                                                  0IGdldCByZXNvdXJjZSAibmFtZXNwYW
creating instance: provider binary /opt/garm/bin/garm-provider-k8s returned error: provider binary failed with stdout: ; stderr: 2023/12/08 12:58:38 failed to run command: failed to create instance in provider: ensuring runner namespace runner failed: namespaces "runner2/08 12:58:38 failed to run com" is forbidden: User "system:serviceaccount:garm-server:garm-server" cannot get resource "namespaces" in API group "" in the namespace rviceaccount:garm-server:garm-s"runner"
: exit status 1
```